src/AppInstallerCLICore/VTSupport.cpp

```diff
@@ -17,92 +17,198 @@ namespace AppInstaller::CLI::VirtualTerminal
-        if (hOut == INVALID_HANDLE_VALUE)
-        {
-            LOG_LAST_ERROR();
-        }
-        else if (hOut == NULL)
+        bool InitializeMode(DWORD handle, DWORD& previousMode, std::initializer_list<DWORD> modeModifierFallbacks, DWORD disabledFlags = 0)
         {
-            AICLI_LOG(CLI, Info, << "VT not enabled due to null output handle");
-        }
-        else
-        {
-            if (!GetConsoleMode(hOut, &m_previousMode))
+            HANDLE hStd = GetStdHandle(handle);
+            if (hStd == INVALID_HANDLE_VALUE)
+            {
+                LOG_LAST_ERROR();
+            }
+            else if (hStd == NULL)
             {
-                // If the user redirects output, the handle will be invalid for this function.
-                // Don't log it in that case.
-                LOG_LAST_ERROR_IF(GetLastError() != ERROR_INVALID_HANDLE);
+                AICLI_LOG(CLI, Info, << "VT not enabled due to null handle [" << handle << "]");
             }
             else
             {
-                // Try to degrade in case DISABLE_NEWLINE_AUTO_RETURN isn't supported.
-                for (DWORD mode : { ENABLE_VIRTUAL_TERMINAL_PROCESSING | DISABLE_NEWLINE_AUTO_RETURN, ENABLE_VIRTUAL_TERMINAL_PROCESSING})
+                if (!GetConsoleMode(hStd, &previousMode))
                 {
-                    DWORD outMode = m_previousMode | mode;
-                    if (!SetConsoleMode(hOut, outMode))
-                    {
-                        // Even if it is a different error, log it and try to carry on.
-                        LOG_LAST_ERROR_IF(GetLastError() != STATUS_INVALID_PARAMETER);
-                    }
-                    else
+                    // If the user redirects output, the handle will be invalid for this function.
+                    // Don't log it in that case.
+                    LOG_LAST_ERROR_IF(GetLastError() != ERROR_INVALID_HANDLE);
+                }
+                else
+                {
+                    for (DWORD mode : modeModifierFallbacks)
                     {
-                        m_token = true;
-                        break;
+                        DWORD outMode = (previousMode & ~disabledFlags) | mode;
+                        if (!SetConsoleMode(hStd, outMode))
+                        {
+                            // Even if it is a different error, log it and try to carry on.
+                            LOG_LAST_ERROR_IF(GetLastError() != STATUS_INVALID_PARAMETER);
+                        }
+                        else
+                        {
+                            return true;
+                        }
                     }
                 }
             }
+
+            return false;
+        }
+
+        // Extracts a VT sequence, expected one of the form ESCAPE + prefix + result + suffix, returning the result part.
+        std::string ExtractSequence(std::istream& inStream, std::string_view prefix, std::string_view suffix)

```

Comment:
If you assume that your suffix is a single character you could use `inStream.getline()` to do all the seeking for you automatically. That would also avoid reading past the end of the DA1 response if you believe that may be a concern.

Personally, at least in the beginning, I'd not care about consuming more input than needed or handling situations where the sequence isn't close to the start of the input buffer, so I'd do something like this:
```cpp
char buf[1024];
const auto len = inStream.readsome(&buf[0], std::size(buf));

std::string_view response{ &buf[0], std::max(0, len) };
const auto beg = response.find('\x1b');
const auto end = response.find('c', beg);

std::string result;
if (beg < end) {
    result.assign(response, beg, end - beg);
}

return result;
```